src/psyclone/psyir/symbols/symbol.py

```diff
@@ -255,6 +255,32 @@ def copy(self):
         return type(self)(self.name, visibility=self.visibility,
                           interface=self.interface)
 
+    def specialise(self, subclass):
+        '''Specialise this class instance so that it becomes an instance of
+        the class provided in the subclass argument. This allows this
+        instance to become a subclass without any references to this
+        instance becoming invalid.
+
+        :param subclass: a subclass of the Symbol class.
+        :type subclass: the type of a class is `type`
+
+        :raises TypeError: if subclass is not a sub-class of Symbol.
+
+        '''
+        try:
+            is_subclass = issubclass(subclass, self.__class__)
+        except TypeError as info:
+            message = ("The specialise method in the Symbol class expects the "
+                       "subclass argument to be a class.")
+            six.raise_from(TypeError(message), info)
+        # pylint: disable = unidiomatic-typecheck
+        if not is_subclass or type(self) is subclass:
+            raise TypeError(
+                "The specialise method in the Symbol class expects the "
+                "subclass argument to be a class that is a sub-class of "
+                "Symbol.")
```

Comment:
"is a sub-class of this Symbol"? Also in docstring

I am also wondering if `type(self) is subclass` is needed. It depends of how we consider a `.specialise(` in the code: as something that definitely happens, or just a guarantee that it is this subclass (if it already is it continues normally). What are your thoughts?

Reply:
I think "of Symbol" is right here as I am talking about the Symbol class of which there is only one, so there is no "this".

It seemed like it would be an error if someone called specialise and did not change the class - as then there is no specialisation. I therefore thought it would be better to report this rather than allow it. Also, I was worried for the case where we allow constructor arguments. It would allow `specialise` to update some of the properties of a class without changing its type. That was my reasoning, but I appreciate there are situations where x is considered to be a subthingy of x and in the end it depends on what is most convenient for the user.

Reply:
You're quite right. I've hopefully updated appropriately.